src/mscorlib/shared/System/IO/PathInternal.cs

```diff
@@ -135,5 +133,20 @@ internal static void CheckSearchPattern(string searchPattern)
                 searchPattern = searchPattern.Substring(index + 2);
             }
         }
+
+        internal static int GetDirectoryNameOffset(ReadOnlySpan<char> path)
+        {
+            int root = GetRootLength(path);
+
+            int i = path.Length;
+            if (i > root)
```

Comment:
Nit: Perhaps you can do an early return by flipping the condition check here. This way can get the invalid case out of the way first. It takes care of error handling right away, saves one level of indentation, and I think looks easier to read.
```
if (i <= root) return -1;
while(i > root && !IsDirectorySeperator(path[--i]));
return i;
```